### DISCUSSION TITLE
Integration with Cookiecutter Django

### DISCUSSION THREAD

**Discussion by timcwilson** [django-commons/django-unicorn#200] [Q&A]
Thanks for django-unicorn Adam, it looks like the best and most painless way to integrate javascript interactions, but I am having trouble integrating with my Cookiecutter project.
With Cookiecutter, the apps are a subdirectory further down under the project directory - something that is alleged to be 'best practice.'
Could you please give some instructions on how to integrate with a non-standard django layout. I keep getting error messages like "Invalid block tag on line 3: 'unicorn'. Did you forget to register or load this tag?"

**Comment by adamghill** [comment]
Interesting! I'm looking at https://github.com/pydanny/cookiecutter-django/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D and I see what you mean.

>Invalid block tag on line 3: 'unicorn'. Did you forget to register or load this tag?

I would think this would be because `django_unicorn` wasn't added to the `INSTALLED_APPS`, but here are a few other things to troubleshoot. Let me know which one you get stuck on and I can help figure out how to make it work.

- Did you add `django-unicorn` to a requirements file and have it installed in the virtual environment? 
- Did you add `django_unicorn` to the `THIRD_PARTY_APPS` setting (like https://github.com/pydanny/cookiecutter-django/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/config/settings/base.py#L72)?
- What happens when you run `python manage.py startunicorn hello-world` from the project directory?

**Reply by timcwilson** [reply]
> Interesting! I'm looking at https://github.com/pydanny/cookiecutter-django/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D and I see what you mean.
> 
> > Invalid block tag on line 3: 'unicorn'. Did you forget to register or load this tag?

I'm not actually sure how to register or load this tag.
My base.py has this:
THIRD_PARTY_APPS = [
    "crispy_forms",
    "django_unicorn",
    "unicorn",
    "allauth",

I've also tried
THIRD_PARTY_APPS = [
    "crispy_forms",
    "django_unicorn",
    "rapid.unicorn", (which is the subdirectory where unicorn is located for me)

But then I've tried copying the unicorn directory to the top level.

> 
> I would think this would be because `django_unicorn` wasn't added to the `INSTALLED_APPS`, but here are a few other things to troubleshoot. Let me know which one you get stuck on and I can help figure out how to make it work.
> 
> * Did you add `django-unicorn` to a requirements file and have it installed in the virtual environment?

Yes, and I ran
pip install django-unicorn

> * Did you add `django_unicorn` to the `THIRD_PARTY_APPS` setting (like https://github.com/pydanny/cookiecutter-django/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/config/settings/base.py#L72)?

yes

> * What happens when you run `python manage.py startunicorn hello-world` from the project directory?

First off it complained because I had no BASE_DIR
I had to add BASE_DIR = APPS_DIR (the sub directory where cookiecutter puts everything)
Once I'd done that, it creates the directory and puts the stubs in the right places.

I've been messing around for a while now, and I get a slightly different error:
Invalid block tag on line 4: 'unicorn_scripts'. Did you forget to register or load this tag?

My html looks like this:
{% extends "base.html" %}
{% load static %}
{% csrf_token %}
{% unicorn_scripts %}
{% unicorn "ufind" %}

**Reply by adamghill** [reply]
>I had to add BASE_DIR = APPS_DIR

Ok I can update the command to handle that better.

Instead of just `{% load static %}` can you do `{% load static unicorn %}`.

If there is anything on https://www.django-unicorn.com/docs/installation/#integrate-unicorn-with-django that is confusing or would be helpful to change let me know.

**Comment by timcwilson** [comment]
Hi Adam,

Thanks for helping. I can get Hello World on my page now. I'll give you
some more feedback in a few days when I've got some more of my project
working, but for now:

The problem with many "instruction/how to" documents is that on the one
hand the examples are too simple, but on the other too complex. What I mean
in this case is that the example needs to fit into a real django project
not just a bare bones django project. So, you know django thoroughly, and
this package thorougly, but people implementing this may not. One of the
problems I had was trying to put the {% load unicorn %} {% unicorn_scripts
%} {% csrf_token %} into my base.html, but using it in one of my other
templte pages that extends base.html. This is possibly obvious to someone
who knows django well, but anti-intuitive to a django beginner. Even though
I watched the video and read the Installation and Components pages, things
were different to my real world application. If you do a demo video with an
django app it may be clearer....even extend the standard django tutorial.
So if the video started with I've just created the django tutorial polls
app and you start your integration from there, you would be simulating
putting unicorn into a semi-real-world app rather than just a bare-bones
minimum.

At the same time, while the django part is too simple, I'm not interested
yet in getting two of the same components on the same form.....so my eyes
ignore component key and drift down to the example component. Maybe split
the Component page into Your first component and Component Details.

Anyway, thanks again for helping and for the project. I'll star it in
github as soon as I get my dynamic search going later today :-).

Regards,
Tim

**Reply by adamghill** [reply]
>I've just created the django tutorial polls app and you start your integration from there

This is good feedback and it totally makes sense to have some instructions with more realistic scenarios. Thanks for the great idea. 👍 

Hopefully, the rest of the implementation has gone ok. :)
